saleor/graphql/attribute/filters.py

```diff
@@ -330,3 +333,295 @@ class Meta:
         filterset_class = AttributeValueWhere
         description = "Where filtering options for attribute values."
         doc_category = DOC_CATEGORY_ATTRIBUTES
+
+
+CONTAINS_TYPING = dict[Literal["contains_any", "contains_all"], list[str]]
+
+
+def filter_by_contains_referenced_object_ids(
+    attr_id: int | None,
+    attr_value: CONTAINS_TYPING,
+    db_connection_name: str,
+    assigned_attr_model: type[AssignedPageAttributeValue],
+    assigned_id_field_name: Literal["page_id"],
+):
+    """Build a filter expression for objects referencing other entities by global IDs.
+
+    Returns a Q expression to filter objects based on their references
+    to other entities (like: variants, products, pages), identified by
+    global IDs.
+
+    - If `contains_all` is provided, only objects that reference all of the
+    specified global IDs will match.
+    - If `contains_any` is provided, objects that reference at least one of
+    the specified global IDs will match.
+    """
+
+    contains_all = attr_value.get("contains_all")
+    contains_any = attr_value.get("contains_any")
+
+    variant_ids = set()
+    product_ids = set()
+    page_ids = set()
+
+    for obj_id in contains_any or contains_all or []:
+        type_, id_ = graphene.Node.from_global_id(obj_id)
+        if type_ == "Page":
+            page_ids.add(id_)
+        elif type_ == "Product":
+            product_ids.add(id_)
+        elif type_ == "ProductVariant":
+            variant_ids.add(id_)
+
+    expression = Q()
+    if contains_all:
+        if page_ids:
+            expression &= filter_by_contains_referenced_pages(
+                attr_id,
+                {"contains_all": list(page_ids)},
+                db_connection_name,
+                identifier_field_name="id",
+                assigned_attr_model=assigned_attr_model,
+                assigned_id_field_name=assigned_id_field_name,
+            )
+        if product_ids:
+            expression &= filter_by_contains_referenced_products(
+                attr_id,
+                {"contains_all": list(product_ids)},
+                db_connection_name,
+                identifier_field_name="id",
+                assigned_attr_model=assigned_attr_model,
+                assigned_id_field_name=assigned_id_field_name,
+            )
+        if variant_ids:
+            expression &= filter_by_contains_referenced_variants(
+                attr_id,
+                {"contains_all": list(variant_ids)},
+                db_connection_name,
+                identifier_field_name="id",
+                assigned_attr_model=assigned_attr_model,
+                assigned_id_field_name=assigned_id_field_name,
+            )
```

Comment:
What about adding a return here? `contains_all` couldn't be mixed with `contains_any`.